server/session_cache_redis.go

```diff
@@ -0,0 +1,187 @@
+package server
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/go-redis/redis/v8"
+	"github.com/gofrs/uuid"
+	"go.uber.org/zap"
+	"time"
+)
+
+const baseKeyStructure = "brs::nakama::%s::%s"
+
+var namespaceUser = "user"
+
+type RedisSessionCache struct {
+	config      Config
+	logger      *zap.Logger
+	client      *redis.Client
+	backupCache SessionCache
+}
+
+// NewRedisSessionCache returns a new SessionCache that uses Redis to cache sessions.
+// It uses config values to connect to Redis and returns errors if connection fails.
+func NewRedisSessionCache(logger *zap.Logger, config Config) (SessionCache, error) {
+	cacheConfig := config.GetCache()
+
+	if len(cacheConfig.CacheName) == 0 || len(cacheConfig.Address) == 0 {
+		return nil, errors.New("redis session cache: not configured")
+	}
+
+	rbdOptions := &redis.Options{
+		Addr:     cacheConfig.Address,
+		Password: cacheConfig.Password,
+		DB:       0,
+	}
+
+	if cacheConfig.EnableTls {
+		rbdOptions.TLSConfig = &tls.Config{
+			MinVersion: tls.VersionTLS12,
+		}
+	}
+
+	redisClient := redis.NewClient(rbdOptions)
+	if redisClient == nil {
+		logger.Error("SessionCache: Not able to get Redis client")
+		return nil, errors.New("redis session cache: not able to get redis connection")
+	} else if _, err := redisClient.Ping(context.Background()).Result(); err != nil {
+		logger.Error("SessionCache: Redis client not properly connected", zap.Error(err))
+		return nil, errors.New("redis session cache: not able to ping redis")
+	} else {
+		logger.Info("SessionCache: Redis connection established.")
+	}
+
+	namespaceUser = fmt.Sprintf(baseKeyStructure, cacheConfig.CacheName, namespaceUser)
+
+	return &RedisSessionCache{
+		config: config,
+		logger: logger,
+		client: redisClient,
+	}, nil
+}
+
+// Stop is called by Nakama when gracefully shutdown is requested
+func (s *RedisSessionCache) Stop() {
+	s.client.Close()
+}
+
+// IsValidSession validates the session token of a user.
+func (s *RedisSessionCache) IsValidSession(userID uuid.UUID, exp int64, token string) bool {
+	cachedUser := s.getCachedUser(userID)
+	if cachedUser == nil {
+		return false
+	}
+
+	return cachedUser.SessionToken == token && cachedUser.SessionExp >= time.Now().UTC().Unix()
+}
+
+// IsValidRefresh validates the refresh token of a user.
+func (s *RedisSessionCache) IsValidRefresh(userID uuid.UUID, exp int64, token string) bool {
+	cachedUser := s.getCachedUser(userID)
+	if cachedUser == nil {
+		return false
+	}
+
+	return cachedUser.RefreshToken == token && cachedUser.RefreshExp >= time.Now().UTC().Unix()
+}
+
+// Add is used when a user should be added to session cache.
+func (s *RedisSessionCache) Add(userID uuid.UUID, sessionExp int64, sessionToken string, refreshExp int64, refreshToken string) {
```

Comment:
Probably a good idea to add a suffix to `refreshExp` describing what type of time value it is - millis, seconds, etc.

Ditto with internal time variables

Reply:
well yes, I was just trying to keep the signature but I'll do the rename